README
Flickr::API
===========

An interface to the Flickr API, including OAuth authentication.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test [ MAKETEST_OAUTH_CFG=config_file, MAKETEST_FLICKR_CFG=other_config, MAKETEST_VALUES=values_file, [MAKETEST_OAUTH_AUTHED=1, MAKETEST_FLICKR_AUTHED=1]]
   make install


NOTES ON API CHANGES

If you use key and secret when instantiating a new API object, be aware that
the Flickr::API silently changes these to api_key and api_secret for use and
configuration export. Since Flickr is (ambivalently) deprecating their own
authentication in favor of OAuth, in time key and secret will become confusing,
is key api_key or consumer_key?

For now, key and secret are silently changed. In future releases there will be
a warning when they are used.


NOTES ON TESTING

Testing authenticated method calls requires a previously authenticated
consumer key and secret (or api key and secret for Flickr's authentication) 
saved in a configuration with the data in Storable.pm format that the tests 
can read. Tests will need either 'read' or 'write' permissions. The tests
do not use delete permission to delete.

If you are testing a previously authenticated configuration, specifying
MAKETEST_OAUTH_AUTHED=1 or MAKETEST_FLICKR_AUTHED=1 will bypass the 
authentication tests.

In addition, testing authentication requires, at minimum, your Flickr api_key
and api_secret; a file with some minimal configuration data in Storable.pm 
format that the tests can read and write; the ability to open a url on Flickr; 
and the ability to retrieve a redirectfrom Flickr. Since there is some user 
interaction required, full authentication testing has to be done manually 
with something like:

make test MAKETEST_OAUTH_CFG=$HOME/oauth-testing.st [TEST_VERBOSE=1]

or

make test MAKETEST_FLICKR_CFG=$HOME/flickr-auth-testing.st [TEST_VERBOSE=1]

or even

make test MAKETEST_OAUTH_CFG=$HOME/oauth-testing.st \
 MAKETEST_FLICKR_CFG=$HOME/flickr-auth-testing.st [TEST_VERBOSE=1]


There are two scripts in the script directory that can be used to build
and verify a Storable.pm format configuration file:

script/flickr_make_stored_config.pl and 
script/flickr_dump_stored_config.pl

When using OAuth, the api_key and api_secret are called the consumer_key
and consumer_secret respectively. When using the deprecated Flickr
authentication, the api_key and api_secret are sometimes called key 
and secret.

It is possible to use authentication testing to produce a configuration
in storable format that can be imported into the Flickr::API for future use.


MORE ON AUTHENTICATION

When using the original Flickr authentication, there are two methods:
web-based and non-web-based. For web-based the developer defines a
callback URL, which is where the Flickr will redirect the session,
along with a frob.

For non-webbased, you need to make a call to flickr.auth.getFrob to
obtain the frob to exchange for a token. The tests only deal with
web-based authentication. In general, since Flickr deprecated their
original authentication in favor of OAuth, more effort is dedicated
to OAuth.

If you are using Flickr's authentication, and you receive an
"Oops! Flickr can't find a valid callback URL." page, then you
are calling an API that is not web-based.


DEPENDENCIES

This module requires these other modules and libraries:

  Carp
  Data::Dumper
  Digest::MD5
  Encode
  Getopt::Long
  HTTP::Request
  HTTP::Request::Common
  HTTP::Response
  LWP::UserAgent
  Net::OAuth
  parent
  Scalar::Util
  Storable
  Term::ReadLine
  URI
  XML::Parser::Lite::Tree
  XML::LibXML::Simple


Testing this module and using the examples additionally require:

  File::Temp
  Test::More
  Test::Script


COPYRIGHT AND LICENSE

Copyright (C) 2004-2013 Cal Henderson <cal@iamcal.com>
License: Perl Artistic License 2.0

OAuth patches and additions
Copyright (C) 2014-2016 Louis B. Moore <lbmoore@cpan.org>
License: Perl Artistic License 2.0




Changes
Changes to Flickr::API

1.28, 2016-10-11
 * Add Flickr::API::Upload
 * Flickr::API Be more consistent using my ($self, $args,...) = @_; and not shift
 * Flickr::API Change from using $options to $args for consistancy
 * Flickr::API Add upload method
 * Flickr::API add api_type method
 * Flickr::API add api_permissions method
 * Flickr::API stop using "return undef;"
 * Flickr::API begin *_silently_* enforcing api_key/api_secret over key/secret
 * Flickr::API::Response Be more consistent using my ($self, $args,...) = @_; and not shift
 * Added MAKETEST_OAUTH_AUTHED=1 and/or MAKETEST_FLICKR_AUTHED=1 for testing to allow testing
   actual calls without re-authenticating
 * Added more info on testing to README
 * changes to tests for authentication to allow bypassing
 * changes to tests to detect if deprecated key/secret become api_key/api_secret 
 * changed script/flickr_dump_stored_config.pl to allow config file to be passed in ARGV
 * changed script/flickr_make_stored_config.pl to shift configs from key/secret to
   api_key/api_secret for Flickr Authentication and warn if key/secret are used.

1.27, 2015-12-02 
 * Get Flickr::API::People into shape for the Tools.
 * Added some methods for asking about a people object's
   calling user and searched user.

1.26, 2015-11-17
 * Removed references to oauth_token and oauth_token_secret since they
   are redundant.
 * Get some spelling consistent
 * Tidy up a test script

1.25, 2015-10-12
 * Realizing that there is benefit to keeping Flickr::API as close to
   perl core as practical and letting Flickr::Tools use Moo and more
   abstraction, stopped moving all of Flickr::Tools into Flickr::API.
*  Moved Person.pm and Types.pm to Flickr::Tools.
*  Bumped version by five to keep it ahead of Tools, at least for
   the time being.

1.20, 2015-10-05
 * moved Cameras.pm, Reflection.pm and People.pm  to new 
   status reporting.
 * Added Person.pm for the flickr.people methods other than finds
 * Added tests for Person.pm
 * Updated the MANIFEST
 * Add tests to 02-oauth_api.t for new script and for private method
 * Added Types.pm to define types for Moo types
 * Consider splitting the Tools back into a separate distribution 
   so Moo isn't required for the base API.

1.19, 2015-09-16
 * adding Flickr::API::People for the flickr.people methods that find
 * adding t/09-people.t to test People module
 * fixed Reflection.pm for argument lists that are only 1 deep.
 * added test to 07-reflection for method that does not have argument
   array, but rather just a single argument 
 * moved flickr deprecated authentication from $api->{flickr} to
  $api->{fauth} to make better use of $api->{flickr} for subclassing.
 * added a number of status reporting methods to API.pm
 * add _propagate_status to Response.pm
 * add script/flickr_make_test_values.pl for more in-depth testing

1.18, 2015-08-12
 * adding Flickr::API::Cameras
 * adding t/08-cameras.t to test Cameras module
 * replaced the STUB on Reflection.pm POD with the intended text
 * Merge in Pull Request from rjbs

1.17, 2015-08-04
 * adding Flickr::API::Reflection
 * adding t/07-reflection.t to test Reflection module
 * Fixed typo per RT-106107
 * moved from @ISA to use parent in Request.pm and Response.pm
 * added call to and method _initialize to API.pm to support
   inheritance in Reflection.pm et. al.
 * make import_storable_config more generic in API.pm
 * make error_message a little prettier in Response.pm
 * invert changes for chronological convenience


1.16, 2015-07-17
 * merged in horsepunchkid's $perms fixes. (API requires more explicit perms arguments)
 * merged in horsepunchkid's fixes in examples/.
 * fleshed out the $perms fixes, changed from warn to carp.
 * Added to API.pm POD to reflect $perms fixes.
 * Added tests to reflect $perms fixes.
 * bumped version in prep for PAUSE upload.
 
1.15
 * Removed some debugging leftovers
 * bumped version


1.14
 * Added 5.008 as minimum Perl version
 * Moving from XML::Simple to XML::LibXML::Simple
 * Remove example from automatically installing in scripts (Request 105426)
 * move dump and make config scripts to bin and made them install-worthy (Request 105426)
 * Clarified licenses in examples (Request 105426)
 * Rename oauth_export_config to export_config and add ability to export
   Flickr's native, if deprecated, configuration. Similar renames and
   reasons for export_storable_config and import_storable_config.
 * Added a hashref of the response content in addition to a tree
 * Added a number of accessor methods to Response.pm
 * Made code a little more "Perl Best Practices"-ish
 * Rename sign_args to _sign_args

1.12, 1.13, 2015-06-09
 * Cleaned up minor bugs found by testers, improved kwalitee
 * Clarified licensing

1.11, 2014-12-27
 * Added OAuth authentication to Flickr::API and Flickr::API::Request
 * Added some examples

1.08, 2013-06-19
 * Various cleanup from Gabor Szabo

1.07, 2013-05-23
 * Added flag to enable handling of native Unicode strings

1.06, 2013-05-11
 * Added `lwpobj` argument, to allow using subclasses of LWP::UserAgent

1.05, 2012-09-12
 * Updated Flickr API urls from http://www.flickr.com/services/ to http://api.flickr.com/services/.

1.04, 2009-08-25
 * re-fix for perl5.6 - just require a newer version of HTTP::Message which supports
   mime-decoding instead.

1.03, 2009-08-23
 * fix for perl5.6 - when HTTP::Message->decoded_content() comes back empty,
   use content() instead.

1.02, 2009-07-31
 * license update for fedora

1.01, 2008-09-30
 * incorrectly numbered the previous release. gah
 * updated tests to work when the flickr api isn't hit-able

0.10, 2008-09-29
 * Work correctly with latest XML::Parser::Lite::Tree (now requires 0.06)
 * Added patch from Flavio Poletti to allow custom api & auth urls


Makefile.PL
use ExtUtils::MakeMaker;

warn <<EOT;





############################___ NOTICE ___##############################
#                                                                      #
#  Some of the api authentication and results tests require some       #
#  information from you. Scripts script/flickr_make_stored_config.pl   #
#  and script/flickr_make_test_values.pl are an easy way to assemble   #
#  this information. Executing either script with the --help flag will #
#  explain the arguments you need for these scripts.                   #
#                                                                      #
#  The authentication tests use either the MAKETEST_OAUTH_CFG or the   #
#  MAKETEST_FLICKR_CFG environment variables or both. The result tests #
#  use the MAKETEST_VALUES environment variable. You can use any of    #
#  environment variables with the others.                              #
#                                                                      #
# If you want to run these tests, answer `No' to the question below    #
# and run the script or build your own config by hand. When you are    #
# ready for "make test" use:                                           #
#    make test MAKETEST_OAUTH_CFG=/the/path/to/above/filename          #
# or                                                                   #
#    make test MAKETEST_FLICKR_CFG=/the/path/to/above/filename         #
# or                                                                   #
#    make test MAKETEST_VALUES=/the/path/to/above/filename             #
# or any combination of the above.                                     #
#                                                                      #
############################___ NOTICE ___##############################

EOT
my $ans = prompt('Do you want to continue?', 'yes');
exit 0 unless $ans =~ /^y(es)?\z/i;

WriteMakefile(
    'NAME'		=> 'Flickr::API',
    'VERSION_FROM'	=> 'lib/Flickr/API.pm',
    'LICENSE'           => 'perl',
    'MIN_PERL_VERSION'  => '5.008',
    'META_MERGE'        => {
        resources => {
            repository  =>  'https://github.com/iamcal/perl-Flickr-API',
        },
    },
    'PREREQ_PM'        => {
        'LWP::UserAgent' => 0,
        'HTTP::Request' => 0,
        'HTTP::Response' => 0,
        'HTTP::Message' => 1.56, # first version of libwww that supports decoded_content()
        'URI' => 1.18,
        'XML::Parser::Lite::Tree' => 0.06,
        'XML::LibXML::Simple' => 0,
        'Digest::MD5' => 0,
        'Getopt::Long' => 1,
        'parent' => 0,
        'Test::More' => 0,
        'Test::Script' => 0,
        'Net::OAuth' => 0,
    },
    'TEST_REQUIRES'  =>  {
        'Test::Script' => 0,
    },
    'EXE_FILES' =>
        [
            'script/flickr_dump_stored_config.pl',
            'script/flickr_make_stored_config.pl',
            'script/flickr_make_test_values.pl',
        ]
    );
